adapter/config/memory/config.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package memory provides an in-memory volatile config store implementation
+package memory
+
+import (
+	"errors"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	"istio.io/pilot/model"
+)
+
+// Make creates an in-memory config store from a config descriptor
+func Make(descriptor model.ConfigDescriptor) model.ConfigStore {
+	out := store{
+		descriptor: descriptor,
+		data:       make(map[string]map[string]proto.Message),
+		revs:       make(map[string]map[string]string),
+	}
+	for _, typ := range descriptor.Types() {
+		out.data[typ] = make(map[string]proto.Message)
+		out.revs[typ] = make(map[string]string)
+	}
+	return &out
+}
+
+type store struct {
+	descriptor model.ConfigDescriptor
+	data       map[string]map[string]proto.Message
+	revs       map[string]map[string]string
+}
+
+func (cr *store) ConfigDescriptor() model.ConfigDescriptor {
+	return cr.descriptor
+}
+
+// Get implements config registry method
+func (cr *store) Get(typ, key string) (proto.Message, bool, string) {
+	_, ok := cr.data[typ]
+	if !ok {
+		return nil, false, ""
+	}
+	val, exists := cr.data[typ][key]
+	return val, exists, cr.revs[typ][key]
+}
+
+// List implements config registry method
+func (cr *store) List(typ string) ([]model.Config, error) {
```

Comment:
We need to return messages together with revisions. Get could also return a single Config and let the client unpack the struct. This is better than `[]revision, []proto.Message`